exo/inference/mlx/sharded_inference_engine.py

```diff
@@ -53,24 +69,84 @@ async def decode(self, shard: Shard, tokens) -> str:
     await self.ensure_shard(shard)
     tokens = await asyncio.get_running_loop().run_in_executor(self.executor, self.tokenizer.decode, tokens)
     return tokens
+
+  async def save_checkpoint(self, shard: Shard, path: str):
+    await self.ensure_shard(shard)
+    await asyncio.get_running_loop().run_in_executor(self.executor, self.model.save_weights, path)
+
+  async def load_checkpoint(self, shard: Shard, path: str):
```

Comment:
Should also be an abstract method in `InferenceEngine`

Reply:
Mmm, this is where I'm not sure. Some inference engines aren't going to support training just by nature. Maybe eventually we'll need a "TrainEngine" that works separately or something, but adding training bits to the base class means it'll make implementing new inference engines that can't support training obnoxious

Reply:
True. we can also just throw a `NotImplementedError` by default and then if you try to train on an `InferenceEngine` that doesn't support it, you see a big fat error that tells you it's not supported